pkg/ui/server.go

```diff
@@ -0,0 +1,282 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ui
+
+import (
+	"context"
+	"fmt"
+	"html/template"
+	"log"
+	"net/http"
+	"strings"
+)
+
+// Pair represents a key value pair used by the select HTML element.
+type Pair struct {
+	Key  string
+	Text string
+}
+
+// Content defines the displayable parts of the token retrieval form.
+type Content struct {
+	UserLabel     string
+	CategoryLabel string
+	ReasonLabel   string
+	TTLLabel      string
+	Categories    []Pair
+	TTLs          []Pair
+}
+
+// FormDetails represents all the input and content used for the token retrievlal form.
+type FormDetails struct {
+	WindowName string
+	Origin     string
+	PageTitle  string
+	Content    Content
+	Category   string
+	Reason     string
+	TTL        string
+	Errors     map[string]string
+}
+
+// SuccessDetails represents the data used for the success page and the postMessage response to the client.
+type SuccessDetails struct {
+	PageTitle    string
+	Token        string
+	TargetOrigin string
+	WindowName   string
+}
+
+// ForbiddenDetails represents the data used for the forbidden page.
+type ForbiddenDetails struct {
+	PageTitle string
+	Message   string
+}
+
+// Server holds the parsed html templates.
+type Server struct {
+	templates map[string]*template.Template
+	allowList []string
+}
+
+var (
+	categories = []string{"explanation", "breakglass"}
+	ttls       = []string{"15", "30", "60", "120", "240"}
+)
+
+// NewServer creates a new HTTP server implementation that will handle
+// rendering the JVS form and parses the go templates.
+func NewServer(ctx context.Context, cfg *ServiceConfig, tmplLocations map[string]string) (*Server, error) {
+	templateMap := make(map[string]*template.Template)
+
+	// Parse templates
+	for key, path := range tmplLocations {
```

Comment:
I left another comment above, but I think we're duplicating work here that the `html/template` package already does for us. This also has the drawback that, once the server has started, it won't pick up any changes unless you restart it.

Architecturally, we need two things here:

1. We need the concept of a "dev mode" where the files are re-read and parsed on each request. Yea, this is a big slower, but it's a nicer dev ex. It can be controlled with an environment variable.

2. In prod mode, we need to use Go's embedded fs functionality to load the files entirely into memory (so we can ship a single static binary).

It's also likely that, longer term, we'll want to inject helper functions or global data into the templates. You can take a look at some of the EN code for inspiration:

- Defining different FSes: https://github.com/google/exposure-notifications-verification-server/blob/main/pkg/render/assets.go
- Rendering HTML: https://github.com/google/exposure-notifications-verification-server/blob/main/pkg/render/html.go
- Generic rendering logic: https://github.com/google/exposure-notifications-verification-server/blob/main/pkg/render/renderer.go

There's a lot of edge cases with rendering (particularly with leaking memory and recovering from panics) which need to be handled carefully.

Reply:
Im going to try to refactor to mimic the ENVS repo. From what I can tell, it seems like `http.NewServeMux` is never used in ENVS. Is that correct?

Reply:
We use a different routing framework, but routing is different from rendering. Mux == multiplexer.

Reply:
Addressed in the latest push. There was a lot to unpack in terms of structure of that code, i think this is in better shape now.